terraform/transform_reference.go

```diff
@@ -206,21 +206,30 @@ func (t *DestroyValueReferenceTransformer) Transform(g *Graph) error {
 	return nil
 }
 
-// PruneUnusedValuesTransformer is s GraphTransformer that removes local and
-// output values which are not referenced in the graph. Since outputs and
-// locals always need to be evaluated, if they reference a resource that is not
-// available in the state the interpolation could fail.
-type PruneUnusedValuesTransformer struct{}
+// PruneUnusedValuesTransformer is a GraphTransformer that removes local,
+// variable, and output values which are not referenced in the graph. If these
+// values reference a resource that is no longer in the state the interpolation
+// could fail.
+type PruneUnusedValuesTransformer struct {
+	Destroy bool
+}
 
 func (t *PruneUnusedValuesTransformer) Transform(g *Graph) error {
-	// this might need multiple runs in order to ensure that pruning a value
-	// doesn't effect a previously checked value.
+	// Pruning a value can effect previously checked edges, so loop until there
+	// are no more changes.
 	for removed := 0; ; removed = 0 {
 		for _, v := range g.Vertices() {
-			switch v.(type) {
-			case *NodeApplyableOutput, *NodeLocal:
+			switch v := v.(type) {
+			case *NodeApplyableOutput:
+				// If we're not certain this is a full destroy, we need to keep any
+				// root module outputs
+				if v.Addr.Module.IsRoot() && !t.Destroy {
+					continue
+				}
+			case *NodeLocal, *NodeApplyableModuleVariable:
```

Comment:
IMHO this isn't any worse than the current interface situation. We've had to fix a number of issues with types missing interface methods, or being giving the wrong methods, and this at least provides direct references to the types we're looking for. We can coordinate cleaning these up (there's a number of direct type references scattered throughout) along with reorganizing the interface structure.

Reply:
My main concern is to keep all of the business logic for a particular node type in one place. Optional interfaces on the node types are currently that "one place", which as you note have their own problems but do at least have consistency: all logic related to a node type is expressed as methods on that node type.

I'm not trying to be prescriptive about any particular solution, but having the logic spread out into all of the individual transformers and evalnodes -- so that we need to study the whole codebase in order to understand how a particular node type behaves -- is the problem I'm worried about here.

Either way, this is not for this PR, so we can worry about this another day. :grinning: